### PR TITLE
Mark job as unstable dynamically

### PR DESCRIPTION
**Pull request by huydhn**
Allow CI jobs to be marked as unstable dynamically.  This use the same mechanism to disable job but with a different issue title `UNSTABLE JOB_NAME`.

The action will output a `is-unstable` flag to let the CI know if the current job it's running is unstable.  This is similar to the way `keep-going` flag is exposed.  Once this is merged, I will follow up with another PR to actually use `is-unstable` flag in CI.

### Testing

* https://github.com/pytorch/pytorch/issues/102297
  * `is-unstable` set https://github.com/pytorch/pytorch/actions/runs/5114544576/jobs/9194921978#step:9:172 
  * Windows CPU jobs are named unstable https://github.com/pytorch/pytorch/actions/runs/5114544576/jobs/9195186715
* https://github.com/pytorch/pytorch/issues/102298
  * `is-unstable set https://github.com/pytorch/pytorch/actions/runs/5114543738/jobs/9195036258#step:11:139 
  * Dynamo jobs are named unstable https://github.com/pytorch/pytorch/actions/runs/5114543738/jobs/9195036258 
* https://github.com/pytorch/pytorch/issues/102299
  * `is-unstable` set https://github.com/pytorch/pytorch/actions/runs/5114544576/jobs/9194922158#step:9:190 
  * MacOS test jobs are named unstable https://github.com/pytorch/pytorch/actions/runs/5114544576/jobs/9195007882
* https://github.com/pytorch/pytorch/issues/102433
  * `is-unstable` set https://github.com/pytorch/pytorch/actions/runs/5114544572/jobs/9198630766#step:13:265
* https://github.com/pytorch/pytorch/issues/102425 (open temporarily during testing)
  * Disabling CI jobs still works correctly https://github.com/pytorch/pytorch/actions/runs/5114543738/jobs/9194904007 (backwards_compat)

### CODE REVIEW

**Comment by huydhn** [comment]
Per the discussion with @clee2000, there will be another part of this PR to update trymerge to handle unstable failures:

* We still want to unstable jobs to fail to gather their signals
* Unstable failure shouldn't block merge

File: .github/scripts/filter_test_configs.py [hunk 1]
```diff
@@ -191,12 +200,86 @@ def set_periodic_modes(
...
+    filtered_test_matrix: Dict[str, List[Any]] = {"include": []}
+
+    # This is an issue to disable a CI job
+    if issue_type == IssueType.DISABLED:
+        # If there is a target config, disable (remove) only that
+        if target_cfg:
+            # Remove the target config from the test matrix
+            filtered_test_matrix["include"] = [
+                r for r in test_matrix["include"] if r.get("config", "") != target_cfg
+            ]
```
Comment (ZainRizvi) [inline .github/scripts/filter_test_configs.py:253]: Do you need a line setting `r[IssueType.DISABLE.value] = IssueType.DISABLE.value`, similar to what UNSTABLE values have below?

Can you please explain why you're tracking the Unstable status? And why the Disabled status might not need to be tracked?

File: .github/scripts/filter_test_configs.py [hunk 2]
```diff
@@ -191,12 +200,86 @@ def set_periodic_modes(
...
+        return filtered_test_matrix
+
+    if issue_type == IssueType.UNSTABLE:
+        for r in test_matrix["include"]:
+            cpy = r.copy()
+
+            if (target_cfg and r.get("config", "") == target_cfg) or not target_cfg:
+                # If there is a target config, only mark that as unstable, otherwise,
+                # mark everything as unstable
+                cpy[IssueType.UNSTABLE.value] = IssueType.UNSTABLE.value
```
Comment (ZainRizvi) [inline .github/scripts/filter_test_configs.py:264]: A dict with `{"unstable": "unstable"}` seems redundant.

What about `{"issue_type": "unstable"}` (which lets you show the "disabled" status and other potential future statuses

**Review by ZainRizvi — APPROVED** [review]

File: .github/scripts/filter_test_configs.py (hunk 2 above)
Comment (huydhn) [reply .github/scripts/filter_test_configs.py:264]: Ah yeah, now that I think about it `{"unstable": "unstable"}` sounds weird.  I copy this convention from ``{"mem_leak_check": "mem_leak_check"}`.  Let me see if `{"issue_type": "unstable"}` works (it should)

File: .github/scripts/filter_test_configs.py (hunk 1 above)
Comment (huydhn) [reply .github/scripts/filter_test_configs.py:253]: This is not for disabled tests as the entry for a disabled job is removed and the job isn't even run.  For example, if I have the following test matrix:

```
{ include: [
  { config: "default", shard: 1, num_shards: 3, runner: "linux.2xlarge" },
  { config: "default", shard: 2, num_shards: 3, runner: "linux.2xlarge" },
  { config: "default", shard: 3, num_shards: 3, runner: "linux.2xlarge" },
  { config: "distributed", shard: 1, num_shards: 2, runner: "linux.2xlarge" },
  { config: "distributed", shard: 2, num_shards: 2, runner: "linux.2xlarge" },
]}
```

where distributed tests are disabled.  The filtered matrix will become:

```
{ include: [
  { config: "default", shard: 1, num_shards: 3, runner: "linux.2xlarge" },
  { config: "default", shard: 2, num_shards: 3, runner: "linux.2xlarge" },
  { config: "default", shard: 3, num_shards: 3, runner: "linux.2xlarge" },
]}
```

On the other hand, if distributed tests are marked as unstable, they would still be run and the filtered matrix will be:

```
{ include: [
  { config: "default", shard: 1, num_shards: 3, runner: "linux.2xlarge" },
  { config: "default", shard: 2, num_shards: 3, runner: "linux.2xlarge" },
  { config: "default", shard: 3, num_shards: 3, runner: "linux.2xlarge" },
  { config: "distributed", shard: 1, num_shards: 2, runner: "linux.2xlarge", issue_type: "unstable" },
  { config: "distributed", shard: 2, num_shards: 2, runner: "linux.2xlarge", issue_type: "unstable" },
]}
```

File: .github/scripts/filter_test_configs.py (hunk 2 above)
Comment (huydhn) [reply .github/scripts/filter_test_configs.py:264]: On second thought, I think this is a minor thing quirk that we can keep as it makes is slightly easier to check if the job has a status, i.e. `"mem_leak_check" in  test_matrix["include"]` or `"unstable" in  test_matrix["include"]`.

**Pull request closed without merging**